FprimeZephyrReference/Components/Authenticate/Authenticate.cpp

```diff
@@ -0,0 +1,532 @@
+    // authenticating or because its not a valid packet.
+    // later we will change the integration tests to all run on plugins, but for now, only the 12 bytes will be
+    // authenticated.
+    if (data.getSize() != 12 + SECURITY_HEADER_LENGTH + SECURITY_TRAILER_LENGTH) {
+        // return the packet, set to unauthenticated
+        U32 newCount = this->rejectedPacketsCount.load() + 1;
+        this->rejectedPacketsCount.store(newCount);
+        this->tlmWrite_RejectedPacketsCount(newCount);
+        this->persistToFile(REJECTED_PACKETS_COUNT_PATH, newCount);
+        this->log_WARNING_HI_PacketTooShort(data.getSize());
+        contextOut.set_authenticated(0);
+        this->dataOut_out(0, data, contextOut);
+        return;
+    }
+
+    // Take the first 6 bytes as the security header
+    unsigned char securityHeader[SECURITY_HEADER_LENGTH];
+    std::memcpy(securityHeader, data.getData(), SECURITY_HEADER_LENGTH);
+    // increment the pointer to the data to point to the rest of the packet
+    data.setData(data.getData() + SECURITY_HEADER_LENGTH);
+    data.setSize(data.getSize() - SECURITY_HEADER_LENGTH);
+
+    // now we get the footer (last 16 bytes)
+    unsigned char securityTrailer[SECURITY_TRAILER_LENGTH];
+    std::memcpy(securityTrailer, data.getData() + data.getSize() - SECURITY_TRAILER_LENGTH, SECURITY_TRAILER_LENGTH);
+    // decrement the size of the data to remove the footer
+    data.setSize(data.getSize() - SECURITY_TRAILER_LENGTH);
+
+    // the first two bytes are the SPI
+    U32 spi = (static_cast<U32>(securityHeader[0]) << 8) | static_cast<U32>(securityHeader[1]);
+
+    // the next four bytes are the sequence number
+    U32 sequenceNumber = (static_cast<U32>(securityHeader[2]) << 24) | (static_cast<U32>(securityHeader[3]) << 16) |
+                         (static_cast<U32>(securityHeader[4]) << 8) | static_cast<U32>(securityHeader[5]);
+
+    const AuthenticationConfig authConfig = this->lookupAuthenticationConfig(spi);
+    const std::string type_authn = authConfig.type;
+    const std::string& key_authn = authConfig.key;
+
+    // check the sequence number is valid
+    // cast trailer to U32
+    bool sequenceNumberValid = this->validateSequenceNumber(sequenceNumber, this->get_SequenceNumber());
+    if (!sequenceNumberValid) {
+        U32 newCount = this->rejectedPacketsCount.load() + 1;
+        this->rejectedPacketsCount.store(newCount);
+        this->tlmWrite_RejectedPacketsCount(newCount);
+        this->persistToFile(REJECTED_PACKETS_COUNT_PATH, newCount);
+        contextOut.set_authenticated(0);
+        this->dataOut_out(0, data, contextOut);
+        return;
+    } else {
+        // increment the stored sequence number
+        U32 newSequenceNumber = sequenceNumber + 1;
+        this->sequenceNumber.store(newSequenceNumber);
+        this->tlmWrite_CurrentSequenceNumber(newSequenceNumber);
+        this->persistToFile(SEQUENCE_NUMBER_PATH, newSequenceNumber);
```

Comment:
I suggest updating the sequence after full packet validation. Is there a good reason to do this work before authenticating the rest of the packet?